### PR TITLE
Add index to local entities

### PR DESCRIPTION
**Pull request by seadowg**
Work towards #5623 

This adds a new `index` property for local entities that can be used as the multiplicity (XPath "index") for each `TreeElement` when we generate a secondary instance. This isn't strictly required right now, but it will be needed later when we want to query `EntitiesRepository` in a `FilterStrategy` to allow for optimized entity filters (ultimately allowing larger entity lists to be used).

#### Why is this the best possible solution? Were any other approaches considered?

The biggest change here was actually switching to a `New`/`Saved` sealed class/ADT pattern for `Entity` to allow for properties that are only available after saving (with `EntitiesRepository`). This allows us to keep the logic of generating `index` entirely within the repository implementations. Any other quirks will be pointed out inline.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

The biggest risk here is that the changes here could mess with forms that use entity lists in selects or calculations. It would be good to play around with those along with deleting (on the server) and adding entities (both on different devices/server and offline).

#### Before submitting this PR, please make sure you have:
- [x] added or modified tests for any new or changed behavior
- [x] run `./gradlew connectedAndroidTest` (or `./gradlew testLab`) and confirmed all checks still pass
- [x] added a comment above any new strings describing it for translators
- [x] added any new strings with date formatting to `DateFormatsTest`
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

### CODE REVIEW

File: collect_app/src/test/java/org/odk/collect/android/entities/EntitiesRepositoryTest.kt
Comment (seadowg) [inline collect_app/src/test/java/org/odk/collect/android/entities/EntitiesRepositoryTest.kt:239]: They don't strictly need to be in insert order most likely, but this behaviour is much easier to test, and will mean the resulting multiplicity within form entry will match the order in the CSVs.

File: collect_app/src/test/java/org/odk/collect/android/entities/EntitiesRepositoryTest.kt
Comment (seadowg) [inline collect_app/src/test/java/org/odk/collect/android/entities/EntitiesRepositoryTest.kt]: This actually might not be what we want, but this drives out actually persisting index (as we don't have a query method to test yet). It also felt important to make the interaction between index values and delete explicit.

Would be good to get a second opinion here @lognaturel.

Comment (seadowg) [reply collect_app/src/test/java/org/odk/collect/android/entities/EntitiesRepositoryTest.kt]: Oh actually this totally won't work: multiplicity must start at 0 and not contain any gaps (with the current version of JavaRosa). I'll mark this as a draft again and add that constraint.

File: collect_app/src/main/java/org/odk/collect/android/entities/JsonFileEntitiesRepository.kt
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/entities/JsonFileEntitiesRepository.kt:19]: Our future "query" APIs would just filter the result of this.

This sort of thing is obviously going to be much more awkward to implement in SQL, but I think we have some options. Worst case, we'll need to do more work in form download (or in some new background job kicked off after form download) to give ourselves an "index".

File: entities/src/main/java/org/odk/collect/entities/storage/Entity.kt
Comment (lognaturel) [inline entities/src/main/java/org/odk/collect/entities/storage/Entity.kt:35]: This is currently just for testing it looks like. It feels like two identical items at different positions in document order should not be considered the same but maybe it doesn't matter for the intended usage.

Is == in Kotlin only going to compare the list references for properties? Maybe that makes my point above moot.

**Review by lognaturel — COMMENTED** [review]
This looks reasonable to me! Just one question/comment.

File: entities/src/main/java/org/odk/collect/entities/storage/Entity.kt
Comment (seadowg) [reply entities/src/main/java/org/odk/collect/entities/storage/Entity.kt:35]: > This is currently just for testing it looks like. It feels like two identical items at different positions in document order should not be considered the same but maybe it doesn't matter for the intended usage.

The idea here was to be able to compare entities without the data that gets added by our repo (currently just `index`, but might also be things like a db ID down the line.). We might have actually end up using this in real code I could imagine.

> Is == in Kotlin only going to compare the list references for properties?

Because `Entity` is a `data class` it'll have an intelligent `equals` implementation (and Kotlin calls `equals` rather than doing reference equality with `==`).

Comment (lognaturel) [reply entities/src/main/java/org/odk/collect/entities/storage/Entity.kt:35]: In the case of Entities, we're (essentially) guaranteed a unique ID so this is fine. In the case of arbitrary secondary instances, I think the index may be important to consider because duplicates are more likely. We can likely come back to it then.

**Review by lognaturel — APPROVED** [review]
I've given enough of a look at this to feel comfortable proceeding. I know @grzesiek2010 is tied up with point release things and getting this in will unblock downstream PRs.

**Comment by grzesiek2010** [comment]
I already had my first pass so if you would like to wait today with merging then I can finish.

File: collect_app/src/test/java/org/odk/collect/android/entities/EntitiesRepositoryTest.kt
Comment (grzesiek2010) [inline collect_app/src/test/java/org/odk/collect/android/entities/EntitiesRepositoryTest.kt:266]: I think in this test it would be good to have more entities and update the first one for example.

File: entities/src/test/java/org/odk/collect/entities/javarosa/LocalEntitiesFileInstanceParserTest.kt
Comment (grzesiek2010) [inline entities/src/test/java/org/odk/collect/entities/javarosa/LocalEntitiesFileInstanceParserTest.kt]: How do we know it's not always 0? It would be good to save more entities and check the multiplicity for them.

**Review by grzesiek2010 — CHANGES_REQUESTED** [review]

**Review by grzesiek2010 — APPROVED** [review]

**Comment by WKobus** [comment]
**Tested with Success!**

Verified on devices with Android 14, 10

**Verified Cases:**
- Adding, updating and deleting entities on server
- Adding and updating entities on different devices in Collect
- Forms with entity list in select
- Offline entities

**Pull request merged**
